File: shared/enrichGameContext.js
//  shared/enrichGameContext.js

import {
  normalizeTeamAbbreviation,
  getTeamInfoByAbbr,
  teamIdMap,
} from "./teamNameMap.js";
import { getGamePkForTeamOnDate } from "../backend/scripts/shared/fetchGameID.js";
import { getGameSchedule } from "./mlbApiUtilsFrontend.js";
import { getTimeOfDayBucketET } from "./timeUtils.js"; // shared-safe

// 🔍 Helper to fetch live feed for pitcher info
async function getFeedLive(gameId) {
  const url = `https://statsapi.mlb.com/api/v1.1/game/${gameId}/feed/live`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch live feed for game ${gameId}`);
  }
  return await res.json();
}

/**
 * Enriches a user-added prop with game context fields:
 * - game_id
 * - is_home
 * - opponent
 * - opponent_encoded
 * - game_time
 * - game_day_of_week
 * - time_of_day_bucket
 * - starting_pitcher_id
 */
export async function enrichGameContext({ team, gameDate }) {
  const normalizedTeam = normalizeTeamAbbreviation(team);

  // 🎮 Resolve game ID
  const gameId = await getGamePkForTeamOnDate(normalizedTeam, gameDate);
  if (!gameId) {
    console.warn(`❌ No game ID found for ${team} on ${gameDate}`);
    return null;
  }

  // 📅 Get full schedule
  const schedule = await getGameSchedule(gameDate);
  const game = schedule.find((g) => g.gamePk === gameId);
  if (!game) {
    console.warn(`❌ Game ID ${gameId} not found in schedule for ${gameDate}`);
    return null;
  }

  // 🏟️ Determine home/away and opponent
  const homeId = game.teams.home.team.id;
  const awayId = game.teams.away.team.id;

  const teamInfo = getTeamInfoByAbbr(normalizedTeam);
  if (!teamInfo) {
    console.warn(`❌ Could not resolve team info for ${normalizedTeam}`);
    return null;
  }

  const isHome = teamInfo.id === homeId;
  const opponentId = isHome ? awayId : homeId;

  // 🧪 Debug logs to trace opponent resolution
  console.log(`🧪 Resolved opponentId: ${opponentId}`);
  console.log(`🧪 teamIdMap[${opponentId}]:`, teamIdMap[opponentId]);

  const opponentInfo = teamIdMap[opponentId];

  console.log(`🧪 Resolved opponentId: ${opponentId}`);
  console.log(`🧪 teamIdMap[${opponentId}]:`, teamIdMap[opponentId]);

  const opponent = opponentInfo?.abbr ?? null;
  const opponent_encoded = opponentId ?? null;

  if (!opponent) {
    console.warn(
      `❌ Could not resolve opponent for ${normalizedTeam} on ${gameDate}`
    );
  }

  // 🕒 Game time context
  const gameTime = game.gameDate;
  const time_of_day_bucket = getTimeOfDayBucketET(gameTime);
  const game_day_of_week = new Date(gameTime).getDay(); // 0 = Sunday

  // 👤 Fetch probable pitcher from live feed
  let starting_pitcher_id = null;
  try {
    const liveFeed = await getFeedLive(gameId);
    const probable = liveFeed?.gameData?.probablePitchers;
    if (isHome) {
      starting_pitcher_id = probable?.away?.id ?? null;
    } else {
      starting_pitcher_id = probable?.home?.id ?? null;
    }
  } catch (err) {
    console.warn(
      `⚠️ Could not fetch starting pitcher for game ${gameId}:`,
      err.message
    );
  }

  const enriched = {
    game_id: gameId,
    is_home: isHome,
    opponent,
    opponent_encoded,
    game_time: gameTime,
    game_day_of_week,
    time_of_day_bucket,
    starting_pitcher_id,
  };

  console.log(`🎯 Enriched game context for ${team} on ${gameDate}:`, enriched);
  return enriched;
}

File: backend/scripts/shared/enrich_game_context.py
# File: backend/scripts/shared/enrich_game_context.py

import requests
from datetime import datetime
from backend.scripts.shared.team_name_map import normalize_team_abbreviation, team_id_map
from backend.scripts.shared.time_utils_backend import get_time_of_day_bucket_et

MLB_API_SCHEDULE = "https://statsapi.mlb.com/api/v1/schedule?sportId=1&date="
MLB_API_FEED = "https://statsapi.mlb.com/api/v1.1/game/{game_id}/feed/live"

def get_schedule(date):
    url = f"{MLB_API_SCHEDULE}{date}"
    res = requests.get(url)
    res.raise_for_status()
    return res.json()

def get_feed_live(game_id):
    url = MLB_API_FEED.format(game_id=game_id)
    res = requests.get(url)
    res.raise_for_status()
    return res.json()

def encode_team(team_abbr):
    teams = list(team_id_map.keys())
    return teams.index(team_abbr) if team_abbr in teams else -1

def enrich_game_context(input):
    player_id = input.get("player_id")
    team_abbr = normalize_team_abbreviation(input.get("team"))
    game_id = input.get("game_id")

    if not game_id:
        raise ValueError("Missing game_id in input")

    feed = get_feed_live(game_id)
    game_data = feed.get("gameData", {})
    datetime_str = game_data.get("datetime", {}).get("dateTime")

    if not datetime_str:
        raise ValueError("Missing game datetime from feed")

    game_time = datetime.fromisoformat(datetime_str.replace("Z", "+00:00"))
    time_of_day_bucket = get_time_of_day_bucket_et(game_time)

    home_team = game_data.get("teams", {}).get("home", {}).get("abbreviation")
    away_team = game_data.get("teams", {}).get("away", {}).get("abbreviation")

    is_home = int(home_team == team_abbr)
    opponent = away_team if is_home else home_team
    opponent_encoded = encode_team(opponent)

    # 🔍 Probable Pitcher
    probable_pitcher_id = None
    try:
        probable = game_data.get("probablePitchers", {})
        pitcher = probable.get("away") if is_home else probable.get("home")
        probable_pitcher_id = pitcher.get("id") if pitcher else None
    except Exception:
        probable_pitcher_id = None

    return {
        "game_id": game_id,
        "game_time": game_time.isoformat(),
        "game_day_of_week": game_time.weekday(),
        "time_of_day_bucket": time_of_day_bucket,
        "is_home": is_home,
        "opponent": opponent,
        "opponent_encoded": opponent_encoded,
        "starting_pitcher_id": probable_pitcher_id,
    }
